opm/simulators/wells/BlackoilWellModel_impl.hpp

```diff
@@ -1241,7 +1245,141 @@ namespace Opm {
+                        auto& ws = well_state.well(well_name);
+                        if (group.hasWell(well_name)) {
+                            well->setDynamicThpLimit(well_group_thp);
+                            const Well& well_ecl = this->wells_ecl_[well->indexOfWell()];
+                            const auto inj_controls = Well::InjectionControls(0);
+                            const auto prod_controls = well_ecl.productionControls(summary_state);
+                            well->iterateWellEqWithSwitching(this->simulator_, dt, inj_controls, prod_controls, well_state, group_state, local_deferredLogger,  false, false);
+                            rate = -tcalc.calcModeRateFromRates(ws.surface_rates);
+                            well_group_rate += rate;
+                        }
+                    }
+                    return (well_group_rate - orig_target)/orig_target;
+                };
 
+                Scalar min_thp, max_thp;
+                std::array<Scalar, 2> range_initial;
+                //Find an initial bracket
+                if (!this->well_group_thp_calc_.has_value()){
+                    // Retrieve the terminal pressure of the associated root of the manifold group
+                    std::string node_name =  nodeName;
+                    while (!network.node(node_name).terminal_pressure().has_value()) {
+                        auto branch = network.uptree_branch(node_name).value();
+                        node_name = branch.uptree_node();
+                    }
+
+                    min_thp = network.node(node_name).terminal_pressure().value();
+                    WellBhpThpCalculator<Scalar>::bruteForceBracketCommonTHP(mismatch, min_thp, max_thp);
+                    // Narrow down the bracket
+                    Scalar low1, high1;
+                    std::array<Scalar, 2> range = {0.9*min_thp, 1.1*max_thp};
+                    std::optional<Scalar> appr_sol;
+                    WellBhpThpCalculator<Scalar>::bruteForceBracketCommonTHP(mismatch, range, low1, high1, appr_sol, 0.0, local_deferredLogger);
+                    min_thp = low1;
+                    max_thp = high1;
+                    range_initial = {min_thp, max_thp};
+                }
+
+                const auto upbranch = network.uptree_branch(nodeName);
+                const auto it = this->node_pressures_.find((*upbranch).uptree_node());
+                const Scalar nodal_pressure = it->second;
+                Scalar well_group_thp = nodal_pressure;
+
+                if (!this->well_group_thp_calc_.has_value() || this->well_group_thp_calc_ > nodal_pressure) {
+                    // The bracket is based on the initial bracket or on a range based on a previous calculated common group thp
+                    std::array<Scalar, 2> range;
+                    this->well_group_thp_calc_.has_value() ? 
+                        range =  {0.9 * this->well_group_thp_calc_.value(), 1.1 * this->well_group_thp_calc_.value()} : 
```

Comment:
This line be problematic for some scenarios. We can discuss to address this one.  Possibly, we can always use `range_initial`.  `range_initial` is already a bracket smaller than 1 bar. Let me know what you think.

Reply:
I understand your concern. However I am failing to make this robust. E.g.  using `range_initial`  did not work. Maybe we can have chat about this.

Reply:
I am just commenting for information.

When the search for `range_initial` works, `range_initial` indeed has proper value and smaller range, 
```
 range_initial 33.781000000000006 33.806666666666672
 range 30.397929434506676 37.153024864397047
 range_initial 33.781000000000006 33.806666666666672
 range 30.397929434506676 37.153024864397047
 range_initial 33.781000000000006 33.806666666666672
```

But sometimes the search for `range_initial` is not always working properly, that will cause problem. 
```
 range_initial 38.500000000000007 38.500000000000007
 range 30.397929434506676 37.153024864397047
 well B-1H gets STOPPED during iteration 
 well B-3H gets STOPPED during iteration 
 range_initial 1123.100000000000364 1123.100000000000364
 range 30.397929434506676 37.153024864397047
 range_initial 1123.100000000000364 1123.100000000000364
 range 30.397929434506676 37.153024864397047
```